pkg/healthchecker/health_checker_windows.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2021 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package healthchecker
+
+import (
+	"fmt"
+	"net/http"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/node-problem-detector/cmd/healthchecker/options"
+	"k8s.io/node-problem-detector/pkg/healthchecker/types"
+	"k8s.io/node-problem-detector/pkg/util"
+)
+
+// getUptimeFunc returns the time for which the given service has been running.
+func getUptimeFunc(service string) func() (time.Duration, error) {
+	return func() (time.Duration, error) {
+		// Using the WinEvent Log Objects to find the Service logs' time when the Service was last stopped/terminated.
```

Comment:
What happens for the first run of the service, and no stopped/terminated log exists?

Reply:
This will only get executed when the the service detects it is unhealthy. So if the service runs and there are no logs that it detects that do not indicate the node is unhealthy, then this will not get executed. In other words, nothing happens yet, until some issue is detected.

Reply:
Service unhealthy != service goes down.

It is possible that docker is not healthy but still running, in that case will this function keep returning error?

Reply:
So long as `docker.exe ps` does not return an error, the uptime function will not get executed. We test the heath of docker by executing docker and running a `ps` command. But, as long as `docker.exe ps` returns an error, then yes, this function will keep returning an error until it is repaired successfully. Does that answer what you are trying to ask?

Reply:
`uptime` would always be 0?

How could it be greater than `hc.coolDownTime`, and will the repair ever happen?

Reply:
I think I understand your original question. The uptime function is looking for when docker either last terminated or last stopped. If for some reason docker fails to execute `docker.exe ps` and is still actually running, healthchecker will indicate that the it is unhealthy and then yes, the uptime function will always be 0 because it did not find when docker actually stopped. Then in that case, the repair won't happen.

Reply:
> If for some reason docker fails to execute docker.exe ps and is still actually running, healthchecker will indicate that the it is unhealthy and then yes, the uptime function will always be 0 because it did not find when docker actually stopped. Then in that case, the repair won't happen.

But we want the repair to happen in that case I think. If the repair won't happen in that case, it seems like a bug to me.

Reply:
I am still not quite sure how I would determine docker has entered that state. Also, I am just porting over the same way it is done in Linux. Would it make sense to create another work item to address this for both linux and windows?

Reply:
The systemd timestamp we use does reflect the service start time:
> The InactiveExitTimestamp tracks when a particular systemd unit transitions from the Inactive to Active state, which can be used to mark the beginning of systemd’s activation of cloud-init.

You should use the service start time instead.

Reply:
Ah, you're right. Let me make those changes to get the last start time instead of the last stop/terminate time

Reply:
I'm not sure if the change addresses the original issue concern, but now, if `docker.exe ps` doesn't work, but it's still running, when the uptime function queries the time, it will look for the last running start time. So, in that case, it should attempt to repair the service.